feat: add logout button to BasicInfo error banner for Transak phone already registered error

amitabh94 opened the pull request:
## **Description**

This PR adds a logout button to the BasicInfo screen's error banner when users encounter Transak error code 2020 (phone number already registered with a different email). This allows users to logout and return to the email input screen to authenticate with the correct email address.

**What is the reason for the change?**
When users authenticate with the wrong email in the Transak deposit flow, if they enter a phone number that's already registered to a different email, they see an error message but have no way to logout and try again with the correct email. This forces them to restart the entire app or navigate to settings to logout manually.

**What is the improvement/solution?**
- Added error code 2020 detection from Transak API response structure
- Added localized error message that extracts and displays the registered email
- Added "Log in with email" button in the error banner that appears only for error code 2020
- Clicking the button clears authentication and navigates back to email input screen
- Reuses existing logout functionality and localization strings
- Added comprehensive test coverage

## **Changelog**

CHANGELOG entry: Added logout button to Transak deposit flow when phone number is already registered

## **Related issues**

Fixes: [TRAM-2799](https://consensyssoftware.atlassian.net/browse/TRAM-2799)

## **Manual testing steps**

### Option 1: Testing with Real Transak (E2E Validation)

```gherkin
Feature: Logout on Transak phone already registered error

  Scenario: user logs out when encountering phone already registered error
    Given user has authenticated with email A in Transak deposit flow
    And user navigates to BasicInfo screen
    And user enters a phone number already registered to email B

    When user submits the form
    Then error banner displays with message "This phone number is already in use by {masked_email}. Log in using this email to continue."
    And "Log in with email" button is visible in the error banner

    When user clicks the "Log in with email" button
    Then user is logged out from Transak
    And user is navigated back to the email input screen
    And user can now enter the correct email to authenticate
```

**Note:** Requires two Transak accounts with the same phone number registered to different emails.

---

### Option 2: Testing with Mock Error (Quick Validation)

**1. Add temporary mock** in `BasicInfo.tsx` around line 212:

```typescript
try {
  // TEMPORARY: Simulate Transak error code 2020
  const mockError = Object.assign(new Error('API Error'), {
    error: {
      errorCode: 2020,
      message: 'This phone number is already registered. It has been used by an account created with k****@pedalsup.com. Login with this email to continue.',
    },
  });
  throw mockError;
  
  // Comment out original code for testing:
  // setLoading(true);
  // const { ssn, ...formDataWithoutSsn } = formData;
  // await postKycForm({ personalDetails: { ... } });
```

**2. Run on simulator:**

```bash
yarn watch
# Press 'i' for iOS or 'a' for Android
```

**3. Test the flow:**
- Navigate to Deposit → Verify Identity → BasicInfo
- Fill in form with any valid data
- Click "Continue"
- ✅ Verify error banner shows: "This phone number is already in use by k****@pedalsup.com. Log in using this email to continue."
- ✅ Verify button shows: "Log in with email"
- Click the button
- ✅ Verify navigation back to email input screen
- ✅ Verify no console errors

**4. Clean up:** Remove the mock code before committing

## **Screenshots/Recordings**

### **Before**
Error banner without logout button - user had no way to logout and try with correct email

<img width="787" height="1704" alt="image" src="https://github.com/user-attachments/assets/58285702-078c-4c00-99b9-d36300e6649e" />


### **After**

**Error Object From Transak** 

<img width="2174" height="1330" alt="image" src="https://github.com/user-attachments/assets/3c175ebd-3df9-4447-8fac-3a3cbcca8e34" />

**Error banner with "Log in with email" button as shown in design mockup**

https://github.com/user-attachments/assets/476ae836-beaf-454b-9806-9350920f3cfb


## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I've included tests if applicable
- [x] I've documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I've applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

[TRAM-2799]: https://consensyssoftware.atlassian.net/browse/TRAM-2799?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Adds detection of Transak error 2020 on BasicInfo, displaying a localized message with the masked email and a "Log in with email" action that logs out and navigates to EnterEmail.
> 
> - **Deposit › Verify Identity (`BasicInfo.tsx`)**:
>   - Detects Transak API `error.errorCode === 2020` (phone already registered).
>     - Extracts masked email from `error.message` and shows localized error `deposit.basic_info.phone_already_registered`.
>     - Displays BannerAlert action button "Log in with email" (`testID: "basic-info-logout-button"`).
>     - On click, calls `logoutFromProvider(false)` and navigates to `...createEnterEmailNavDetails()`.
>   - Resets error/UI flags on retry and field changes; integrates `logoutFromProvider` from `useDepositSDK`.
> - **Tests (`BasicInfo.test.tsx`)**:
>   - Adds cases for: showing logout button on `2020` errors, formatting message with email, no action for generic errors, successful logout + navigation, and graceful logout failure logging.
> - **Localization (`locales/languages/en.json`)**:
>   - Adds `deposit.basic_info.phone_already_registered` and `deposit.basic_info.login_with_email`; keeps `unexpected_error`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 6bf83360bab287128a95ef1ac60f9fb783a19a08. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

amitabh94 committed:
feat: add logout button to BasicInfo error banner for Transak phone already registered error

- Add helper function to detect error code 2020 (phone already registered)
- Add logout handler that calls logoutFromProvider and navigates to EnterEmail
- Update error banner to show logout button via actionButtonProps
- Reuse existing localization string from ConfigurationModal
- Add comprehensive test coverage for logout button functionality

Fixes error recovery when user authenticates with wrong email and phone is already registered

amitabh94 committed:
feat: check errorCode 2020 and format error message for phone already registered

- Add isPhoneRegisteredError flag to detect error code 2020
- Check error.errorCode === 2020 from Transak API response structure
- Extract email from error message and format custom message
- Display: 'This phone number is already in use by {email}. Log in using this email to continue.'
- Keep existing error state pattern (string | null) for consistency

amitabh94 committed:
test: update BasicInfo tests for errorCode 2020 handling

- Mock proper Transak API error structure with errorCode 2020
- Test formatted error message display
- Update all test cases to use correct error structure
- Verify email extraction and formatting in error message

amitabh94 committed:
feat: add localization for phone already registered error message

- Add deposit.basic_info.phone_already_registered localization string
- Use strings() with email interpolation for formatted message
- Update tests to verify localized error message
- Extract email from Transak error message for display

amitabh94 committed:
fix: email regex to avoid trailing period and update tests

- Fix email extraction regex to not capture trailing punctuation
- Changed from /[\w*]+@[\w*.]+/ to /[\w*]+@[\w*]+(?:\.[\w*]+)*/
- Add SSN input to all logout tests (US region requires SSN)
- Use testID instead of getByRole for Continue button
- Wait for error messages to appear before asserting logout button
- All 18 tests now passing

amitabh94 committed:
feat: change logout button to 'Login with {email}'

- Add deposit.basic_info.login_with_email localization string
- Store registeredEmail in state when error code 2020 occurs
- Update button label to use email: 'Login with k****@pedalsup.com'
- Clear registeredEmail when errors are cleared or form fields change
- Update tests to verify new button label
- All 18 tests passing

amitabh94 committed:
feat: update logout button label to 'Log in with email'

- Change button label from 'Login with {email}' to 'Log in with email'
- Update localization string to static text without email interpolation
- Update all tests to verify new button label
- All 18 tests passing

wachunei commented on app/components/UI/Ramp/Deposit/Views/BasicInfo/BasicInfo.tsx:
```diff
@@ -227,11 +232,35 @@ const BasicInfo = (): JSX.Element => {
+        errorMessage = email
+          ? strings('deposit.basic_info.phone_already_registered', { email })
+          : errorWithCode.error.message;
+      } else {
```
In the case the error message changes from Transak and we can't read the email anymore using this code we will have this situation:

- we know the error code is `2020` and it must offer the user a way to log in again
- `isPhoneRegisteredError` is `true`
- `registeredEmail` is `''` (this is the case we are assuming)

Later in the `BannerAlert`, `actionButtonProps` will be `undefined` because the condition to set the props (`isPhoneRegisteredError && registeredEmail`) will be `false` (`Boolean(true && '')`), however we know the error is `2020` and must show the log in button. 

I suggest we show the log in button even if `registeredEmail` is `''` since we know the error code is `2020`.

what do you think?

amitabh94 replied:
I think that is a good idea. I was earlier setting the email in the CTA itself but content came back and mentioned that we can do away with just saying 'Login with email'. I have removed the registeredEmail altogether as we don't need it.

wachunei commented on app/components/UI/Ramp/Deposit/Views/BasicInfo/BasicInfo.tsx:
```diff
@@ -240,17 +269,32 @@ const BasicInfo = (): JSX.Element => {
+      await logoutFromProvider(false); // false = no server invalidation needed
+
+      // Navigate back to email entry screen
+      navigation.navigate(...createEnterEmailNavDetails({}));
```
No need to pass empty params.
```suggestion
      navigation.navigate(...createEnterEmailNavDetails());
```

amitabh94 committed:
fix: address Pedro's review comments

- Remove labelTextVariant from logout button (button text should not be smaller than banner text)
- Remove empty params from createEnterEmailNavDetails() call
- Show logout button even if registeredEmail is empty (error code 2020 is sufficient)
- Remove unused registeredEmail state variable
- Update test to expect undefined params instead of empty object
- All 18 tests passing

wachunei commented on app/components/UI/Ramp/Deposit/Views/BasicInfo/BasicInfo.tsx:
```diff
@@ -227,11 +230,34 @@ const BasicInfo = (): JSX.Element => {
+        // Extract email from message like "...created with k****@pedalsup.com..."
+        const emailMatch = errorWithCode.error.message.match(
+          /[\w*]+@[\w*]+(?:\.[\w*]+)*/,
+        );
```
@amitabh94 do we know what obfuscation rule is done on Transaks end? for example, is there always a char followed by `*` or is there always a domain?

just trying to think of edge cases here, probably if the email was delivered in a separate property it would be easier for everyone.

amitabh94 committed:
feat: add labelTextVariant BodyMD to logout button

- Set labelTextVariant to TextVariant.BodyMD for consistent text sizing
- All tests passing, no snapshot updates required

amitabh94 committed:
Merge branch 'main' into feat/transak-logout-on-phone-error

amitabh94 committed:
Merge branch 'main' into feat/transak-logout-on-phone-error

amitabh94 committed:
Merge branch 'main' into feat/transak-logout-on-phone-error
